src/otlp.zig

```diff
@@ -0,0 +1,744 @@
+        }
+        const value: []const u8 = if (kv.next()) |t| std.mem.trim(u8, t, " ") else return ConfigError.InvalidHeadersSyntax;
+        if (value.len == 0) {
+            return ConfigError.InvalidHeadersSyntax;
+        }
+        if (kv.next()) |_| {
+            return ConfigError.InvalidHeadersSyntax;
+        }
+        headers[idx] = std.http.Header{ .name = key, .value = value };
+        idx += 1;
+        // Fail when the sum of all bytes for the headers overflows.
+        // Each header is accompanied by 3 more bytes: a colon, a space and a newline.
+        cum_bytes = std.math.add(u13, cum_bytes, @intCast(key.len + value.len + 3)) catch return ConfigError.InvalidHeadersTooManyBytes;
+    }
+    return headers[0..idx];
+}
+
+test "otlp config parse headers" {
+    const valid_headers = "a=b,123=456,key1=value1  ,  key2=value2";
+    const parsed = try parseHeaders(valid_headers);
+
+    try std.testing.expectEqual(parsed.len, 4);
+    try std.testing.expectEqualSlices(u8, parsed[0].name, "a");
+    try std.testing.expectEqualSlices(u8, parsed[0].value, "b");
+    try std.testing.expectEqualSlices(u8, parsed[1].name, "123");
+    try std.testing.expectEqualSlices(u8, parsed[1].value, "456");
+    try std.testing.expectEqualSlices(u8, parsed[2].name, "key1");
+    try std.testing.expectEqualSlices(u8, parsed[2].value, "value1");
+    try std.testing.expectEqualSlices(u8, parsed[3].name, "key2");
+    try std.testing.expectEqualSlices(u8, parsed[3].value, "value2");
+
+    const invalid_headers: [4][]const u8 = .{ "a=,", "=b", "a=b=c", "a=b,=c=d" };
+    for (invalid_headers) |header| {
+        try std.testing.expectError(ConfigError.InvalidHeadersSyntax, parseHeaders(header));
+    }
+
+    // 150 bytes * 60 == 9000 bytes
+    const invalid_too_many_bytes: []const u8 = "key=AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA," ** 60;
+    try std.testing.expectError(ConfigError.InvalidHeadersTooManyBytes, parseHeaders(invalid_too_many_bytes));
+
+    const invalid_too_many_items: []const u8 = "key=val," ** 65;
+    try std.testing.expectError(ConfigError.InvalidHeadersTooManyItems, parseHeaders(invalid_too_many_items));
+}
+
+// Implements the Exponential Backoff Retry strategy for HTTP requests.
+// The default configuration values are not dictated by the OTLP spec.
+// We choose toset a maxumum number of retries, even if not specified in the spec,
+// to avoid infinite loops in case of a un-responsive server.
+const ExpBackoffRetry = struct {
+    const Self = @This();
+
+    allocator: std.mem.Allocator,
+    client: *http.Client,
+
+    max_retries: u32,
+    base_delay_ms: u64,
+    max_delay_ms: u64,
+
+    queue: std.PriorityQueue(FetchRequest, void, compareFetchRequest),
+    thread: ?std.Thread,
+    mutex: std.Thread.Mutex,
+    condition: std.Thread.Condition,
+    running: std.atomic.Value(bool),
```

Comment:
I tried this to have a more structured approach to retries.
Turns out that it's too complex, and I'm ditching it in #47 tosimply fire off a new thread for each request that needs to be retried.

Alas, you don't _have_ to review this part.